src/_pytest/cacheprovider.py

```diff
@@ -5,39 +5,41 @@
 ignores the external pytest-cache
 """
 from __future__ import absolute_import, division, print_function
-
 from collections import OrderedDict
 
 import py
 import six
+import attr
 
 import pytest
 import json
-import os
 from os.path import sep as _sep, altsep as _altsep
+import shutil
+
+from . import paths
+
+import logging
 
 
+log = logging.getLogger(__name__)
+
+
+@attr.s
 class Cache(object):
 
-    def __init__(self, config):
-        self.config = config
-        self._cachedir = Cache.cache_dir_from_config(config)
-        self.trace = config.trace.root.get("cache")
-        if config.getoption("cacheclear"):
-            self.trace("clearing cachedir")
-            if self._cachedir.check():
-                self._cachedir.remove()
-            self._cachedir.mkdir()
+    @classmethod
+    def for_config(cls, config):
+        cachedir = cls.cache_dir_from_config(config)
+        if config.getoption("cacheclear") and cachedir.exists():
+            shutil.rmtree(str(cachedir))
+            cachedir.mkdir()
+        return cls(cachedir)
+
+    _cachedir = attr.ib(repr=False)
```

Comment:
For consistency we maybe should keep all `attr.ib` declarations at the start of the class declaration, before any methods?